src/Silverback.Core/Messaging/Subscribers/ReturnValueHandlers/AsyncEnumerableMessagesReturnValueHandler.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) 2020 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Silverback.Messaging.Configuration;
+using Silverback.Messaging.Publishing;
+using Silverback.Util;
+
+namespace Silverback.Messaging.Subscribers.ReturnValueHandlers
+{
+    /// <summary>
+    ///     Handles the returned <see cref="IEnumerable{T}" /> republishing all the messages.
+    /// </summary>
+    public class AsyncEnumerableMessagesReturnValueHandler : IReturnValueHandler
+    {
+        private readonly IBusOptions _busOptions;
+
+        private readonly IPublisher _publisher;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="AsyncEnumerableMessagesReturnValueHandler" /> class.
+        /// </summary>
+        /// <param name="publisher">
+        ///     The <see cref="IPublisher" /> to be used to publish the messages.
+        /// </param>
+        /// <param name="busOptions">
+        ///     The <see cref="IBusOptions" /> that specify which message types have to be handled.
+        /// </param>
+        public AsyncEnumerableMessagesReturnValueHandler(IPublisher publisher, IBusOptions busOptions)
+        {
+            _publisher = publisher;
+            _busOptions = busOptions;
+        }
+
+        /// <inheritdoc cref="IReturnValueHandler.CanHandle" />
+        public bool CanHandle(object returnValue) =>
+            returnValue != null &&
+            returnValue.GetType().GetInterfaces().Any(
+                i => i.IsGenericType &&
+                     i.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>) &&
+                     _busOptions.MessageTypes.Any(
+                         messageType =>
+                             messageType.IsAssignableFrom(i.GenericTypeArguments[0])));
+
+        /// <inheritdoc cref="IReturnValueHandler.Handle" />
+        public void Handle(object returnValue)
+        {
+            throw new InvalidOperationException("Shouldn't enter here for any reason...");

```

Comment:
@BEagle1984 how do we handle this specific case?

Here the messages _theoretically_ never arrives, since when you return an IAsyncEnumerable the method *must* be async, therefore it passes only through the HandleAsync method in the ReturnValueHandler.

Do we leave it simply empty instead of throwing the exception? Otherwise, which exception do you advice to throw?